### PR TITLE
Abstract Base Class for Bio/File.py

### PR DESCRIPTION
**Pull request by chris-rands**
This is a test case for the use of ABCs, it is spun out from https://github.com/biopython/biopython/pull/2717#issuecomment-587536727

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Tests/test_File.py [hunk 1]
```diff
@@ -56,14 +56,20 @@ def test_handle(self):
...
             with File.as_handle(fp) as handle:
-                self.assertEqual(fp, handle, "as_handle should "
-                                 "return argument when given a "
-                                 "file-like object")
+                self.assertEqual(
+                    fp,
+                    handle,
+                    "as_handle should "
+                    "return argument when given a "
+                    "file-like object",
```
Comment (peterjc) [inline Tests/test_File.py]: Those strings could be merged to save a line or two

File: Tests/test_File.py [hunk 2]
```diff
@@ -56,14 +56,20 @@ def test_handle(self):
...
                 self.assertFalse(handle.closed)
 
-            self.assertFalse(handle.closed,
-                             "Exiting as_handle given a file-like object "
-                             "should not close the file")
+            self.assertFalse(
+                handle.closed,
+                "Exiting as_handle given a file-like object "
+                "should not close the file",
+            )
```
Comment (peterjc) [inline Tests/test_File.py]: I think that will just fit with the string literal on one line.

**Comment by peterjc** [comment]
Seems OK, but I need to know more about ABCs to be sure. @mdehoon or @MarkusPiotrowski could you comment?

**Comment by chris-rands** [comment]
Thanks Peter, i agree another pair of eyes to review this would be good, since the sub-classing is pretty extensive (BTW, the minor formatting changes were only incidental since I ran `black`)

**Comment by peterjc** [comment]
@mr-c Thank you.

@chris-rands If you could deal with the (black related) conflicts in the tests, then I think we can merge this.

**Comment by chris-rands** [comment]
Thanks @mr-c for the 2 approvals

@peterjc  I fixed the black conflict I think

**Comment by chris-rands** [comment]
I think this can be safely merged if the test failures unrelated?

**Comment by peterjc** [comment]
Yeah, we fixed the broken API build on TravisCI on the master branch.

**Comment by peterjc** [comment]
Thanks all.

**Pull request merged**